lib/theme_check/checks/pagination_size.rb

```diff
@@ -0,0 +1,65 @@
+# frozen_string_literal: true
+module ThemeCheck
+  class PaginationSize < LiquidCheck
+    severity :suggestion
+    categories :performance
+    doc docs_url(__FILE__)
+
+    attr_reader :min_size
+    attr_reader :max_size
+
+    def initialize(min_size: 1, max_size: 50)
+      @min_size = min_size
+      @max_size = max_size
+    end
+
+    def on_document(_node)
+      @paginations = {}
+      @schema_settings = {}
+    end
+
+    def on_paginate(node)
+      size = node.value.page_size
+      unless @paginations.key?(size)
+        @paginations[size] = []
+      end
+      @paginations[size].push(node)
+    end
+
+    def on_schema(node)
+      schema = JSON.parse(node.value.nodelist.join)
+
+      if (settings = schema["settings"])
+        @schema_settings = settings
+      end
+    rescue JSON::ParserError
+      # Ignored, handled in ValidSchema.
+    end
+
+    def after_document(_node)
+      @paginations.each_pair do |size, nodes|
+        numerical_size = if size.is_a?(Numeric)
+          size
+        else
+          get_setting_default_value(size.lookups.last)
+        end
+        if numerical_size.nil?
+          nodes.each { |node| add_offense("Default pagination size should be defined in the section settings", node: node) }
+        elsif numerical_size > @max_size || numerical_size < @min_size || !numerical_size.is_a?(Integer)
+          nodes.each { |node| add_offense("Pagination size must be a positive integer between #{@min_size} and #{@max_size}", node: node) }
+        end
```

Comment:
very nice.

1 more tiny thing: now that we're explicitly testing for the number being an integer, we should also have a test for the case that that numerical_size is in the right range, but not an integer.  in that case it should also add the offence.

Reply:
@taylorsteiger-rad I meant that we needed to add a test in the test suite for this - but it looks like you did this already 👍

Reply:
Ah, I see, that makes sense. Thanks for improving test coverage!